# **Descriptive Analysis**

In [28]:
import pandas as pd
from google.colab import files
import io

file_path = 'ACME-HappinessSurvey2020.csv'
df = pd.read_csv(file_path)

print(f"File '{file_path}' loaded into DataFrame.")

display(df.head())
display(df.tail())

print("""
Target Variable (Y):
  0 = Unhappy customer
  1 = Happy customer

Features (X):
  X1 = My order was delivered on time
  X2 = Contents of my order were as expected
  X3 = I ordered everything I wanted to order
  X4 = I paid a good price for my order
  X5 = I am satisfied with my courier
  X6 = The app makes ordering easy for me
""")

File 'ACME-HappinessSurvey2020.csv' loaded into DataFrame.


,Y,X1,X2,X3,X4,X5,X6
0,0,3,3,3,4,2,4
1,0,3,2,3,5,4,3
2,1,5,3,3,3,3,5
3,0,5,4,3,3,3,5
4,0,5,4,3,3,3,5


,Y,X1,X2,X3,X4,X5,X6
121,1,5,2,3,4,4,3
122,1,5,2,3,4,2,5
123,1,5,3,3,4,4,5
124,0,4,3,3,4,4,5
125,0,5,3,2,5,5,5



Target Variable (Y):
  0 = Unhappy customer
  1 = Happy customer

Features (X):
  X1 = My order was delivered on time
  X2 = Contents of my order were as expected
  X3 = I ordered everything I wanted to order
  X4 = I paid a good price for my order
  X5 = I am satisfied with my courier
  X6 = The app makes ordering easy for me



In [12]:
print('descriptive statistics:\n', df['Y'].describe())
print('\nThere is a high level of dissatisfaction')
print('\ndescriptive statistics:\n', df['X1'].describe())
print('\ndescriptive statistics:\n', df['X2'].describe())
print('\ndescriptive statistics:\n', df['X3'].describe())
print('\ndescriptive statistics:\n', df['X4'].describe())
print('\ndescriptive statistics:\n', df['X5'].describe())
print('\ndescriptive statistics:\n', df['X6'].describe())

descriptive statistics:
 count    126.000000
mean       0.547619
std        0.499714
min        0.000000
25%        0.000000
50%        1.000000
75%        1.000000
max        1.000000
Name: Y, dtype: float64

There is a high level of dissatisfaction

descriptive statistics:
 count    126.000000
mean       4.333333
std        0.800000
min        1.000000
25%        4.000000
50%        5.000000
75%        5.000000
max        5.000000
Name: X1, dtype: float64

descriptive statistics:
 count    126.000000
mean       2.531746
std        1.114892
min        1.000000
25%        2.000000
50%        3.000000
75%        3.000000
max        5.000000
Name: X2, dtype: float64

descriptive statistics:
 count    126.000000
mean       3.309524
std        1.023440
min        1.000000
25%        3.000000
50%        3.000000
75%        4.000000
max        5.000000
Name: X3, dtype: float64

descriptive statistics:
 count    126.000000
mean       3.746032
std        0.875776
min        1.000000
25%       

# **Correlation Matrix and VIF**

In [13]:
from scipy import stats
corr = df.corr(method = 'pearson')
corr.style.background_gradient(cmap='coolwarm')

print("\n No")

,Y,X1,X2,X3,X4,X5,X6
Y,1.000000,0.280160,-0.024274,0.150838,0.064415,0.224522,0.167669
X1,0.280160,1.000000,0.059797,0.283358,0.087541,0.432772,0.411873
X2,-0.024274,0.059797,1.000000,0.184129,0.114838,0.039996,-0.062205
X3,0.150838,0.283358,0.184129,1.000000,0.302618,0.358397,0.203750
X4,0.064415,0.087541,0.114838,0.302618,1.000000,0.293115,0.215888
X5,0.224522,0.432772,0.039996,0.358397,0.293115,1.000000,0.320195
X6,0.167669,0.411873,-0.062205,0.203750,0.215888,0.320195,1.000000


In [16]:
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
import pandas as pd

X = df[['X1','X2','X3','X4','X5','X6']]
X_const = sm.add_constant(X)

vif = pd.DataFrame()
vif['feature'] = X_const.columns
vif['VIF'] = [variance_inflation_factor(X_const.values, i) for i in range(X_const.shape[1])]
print(vif)

  feature        VIF
0   const  56.076264
1      X1   1.421675
2      X2   1.058062
3      X3   1.274234
4      X4   1.194403
5      X5   1.419828
6      X6   1.293416


In [17]:
from scipy import stats

for col in ['X1','X2','X3','X4','X5','X6']:
    happy = df[df.Y==1][col]
    unhappy = df[df.Y==0][col]
    b_stat, b_p = stats.bartlett(happy, unhappy)
    equal_var = b_p > 0.05
    t, p = stats.ttest_ind(happy, unhappy, equal_var=equal_var)
    print(f"{col}: bartlett p={b_p:.3f} (equal_var={equal_var}), t={t:.2f}, p={p:.4f}")

X1: bartlett p=0.584 (equal_var=True), t=3.25, p=0.0015
X2: bartlett p=0.415 (equal_var=True), t=-0.27, p=0.7873
X3: bartlett p=nan (equal_var=False), t=1.70, p=0.0914
X4: bartlett p=nan (equal_var=False), t=0.72, p=0.4711
X5: bartlett p=0.228 (equal_var=True), t=2.57, p=0.0115
X6: bartlett p=nan (equal_var=False), t=1.88, p=0.0629


# **Factor Analysis**

In [19]:
!pip install factor_analyzer

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.8/42.8 kB 568.5 kB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for factor_analyzer: filename=factor_analyzer-0.5.1-py2.py3-none-any.whl size=42655 sha256=a9c107fdd33e255478abb320f3a473ff3f3984bf26ac4f54cfa03e0367c21d09
  Stored in directory: /root/.cache/pip/wheels/a2/af/06/f4d4ed4d9d714fda437fb1583629417319603c2266e7b233cc
Successfully built factor_analyzer


In [21]:
import pandas as pd
from factor_analyzer import FactorAnalyzer

X = df[['X1','X2','X3','X4','X5','X6']]

fa = FactorAnalyzer(method='principal',n_factors=6,rotation='varimax').fit(X)
ev, v = fa.get_eigenvalues()
print('\n Eigan Value :\n', pd.DataFrame(ev))


 Eigan Value :
           0
0  2.199925
1  1.136176
2  0.884735
3  0.701205
4  0.606852
5  0.471106


/usr/local/lib/python3.12/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


In [25]:
fa = FactorAnalyzer(n_factors=2, rotation='varimax')
fa.fit(X)

print("Loadings:\n", pd.DataFrame(fa.loadings_, index=X.columns, columns=['F1','F2']))
print("\nCommunalities:\n", pd.DataFrame(fa.get_communalities(), index=X.columns, columns=['comm']))

print('\nX1: 0.46, X3: 0.44, X5: 0.43, X6: 0.36 are significant\n')

Loadings:
           F1        F2
X1  0.669903  0.087683
X2 -0.047279  0.314422
X3  0.317565  0.582596
X4  0.214370  0.416366
X5  0.572503  0.319549
X6  0.598594  0.038145

Communalities:
         comm
X1  0.456459
X2  0.101097
X3  0.440266
X4  0.219316
X5  0.429872
X6  0.359770

X1: 0.46, X3: 0.44, X5: 0.43, X6: 0.36 are significant



/usr/local/lib/python3.12/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


# **First Summary**

## Summary: Exploratory Analysis

**Dataset:** 126 customers, 6 features (X1–X6, Likert 1–5), binary target Y.
Class balance: 55% happy / 45% unhappy → baseline accuracy = 55%.

### Descriptive Statistics
- X1 (on-time delivery): mean 4.33
- X2 (contents as expected): mean 2.53
- X3–X6: mid-range means

### Correlation with Y (Pearson)
- Weak correlation: X1 (0.28), X5 (0.22), X6 (0.17), X3 (0.15)
- No correlation: X4 (0.06), X2 (-0.02)

### Multicollinearity (VIF, with constant)
All features VIF between 1.05–1.42 so no multicollinearity. All 6 features safe to use.

### t-tests (happy vs unhappy, Bartlett-adjusted)
- Significant (p < 0.05): **X1 (p=0.0015), X5 (p=0.0115)**
- Borderline: X3 (p=0.09), X6 (p=0.06)
- Not significant: X2 (p=0.79), X4 (p=0.47)

### Factor Analysis (Kaiser rule: 2 factors)
- **Factor 1 — Service experience:** X1, X5, X6
- **Factor 2 — Order content/value:** X3, X4, X2
- Communalities: X2 = 0.10 (very poor fit), X4 = 0.22 (poor fit)

### Conclusions
- **Strongest predictors:** X1, X5, X6
- **Drop candidates:** X2 (fails correlation, t-test, and communality) and X4 (fails correlation and t-test, weak communality)
- **X2 is kept as it is borderline**

# **Regression**

In [39]:
import statsmodels.api as sm
import numpy as np

X = df[['X1','X2','X3','X4','X5','X6']]
y = df['Y']
X = sm.add_constant(X)

logit = sm.Logit(y, X).fit()
print(logit.summary2())
print('\nOdds ratios:\n', np.exp(logit.params))

Optimization terminated successfully.
         Current function value: 0.637875
         Iterations 5
                        Results: Logit
Model:              Logit            Method:           MLE     
Dependent Variable: Y                Pseudo R-squared: 0.074   
Date:               2026-05-04 07:04 AIC:              174.7445
No. Observations:   126              BIC:              194.5985
Df Model:           6                Log-Likelihood:   -80.372 
Df Residuals:       119              LL-Null:          -86.764 
Converged:          1.0000           LLR p-value:      0.046597
No. Iterations:     5.0000           Scale:            1.0000  
-----------------------------------------------------------------
         Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
-----------------------------------------------------------------
const   -3.5674     1.4891   -2.3957   0.0166   -6.4859   -0.6489
X1       0.5738     0.2888    1.9866   0.0470    0.0077    1.1400
X2      -0.0937  

In [38]:
from sklearn.metrics import accuracy_score, confusion_matrix
import statsmodels.api as sm

X = df[['X1','X3','X5','X6']]
y = df['Y']
X = sm.add_constant(X)

logit = sm.Logit(y, X).fit()
print(logit.summary2())

pred = (logit.predict(X) > 0.5).astype(int)
print('Accuracy:', accuracy_score(y, pred))
print('Confusion:\n', confusion_matrix(y, pred))

Optimization terminated successfully.
         Current function value: 0.639020
         Iterations 5
                        Results: Logit
Model:              Logit            Method:           MLE     
Dependent Variable: Y                Pseudo R-squared: 0.072   
Date:               2026-05-04 07:04 AIC:              171.0330
No. Observations:   126              BIC:              185.2144
Df Model:           4                Log-Likelihood:   -80.516 
Df Residuals:       121              LL-Null:          -86.764 
Converged:          1.0000           LLR p-value:      0.014023
No. Iterations:     5.0000           Scale:            1.0000  
-----------------------------------------------------------------
         Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
-----------------------------------------------------------------
const   -3.7949     1.3288   -2.8559   0.0043   -6.3993   -1.1905
X1       0.5677     0.2858    1.9862   0.0470    0.0075    1.1280
X3       0.1106  

In [37]:
X = df[['X1', 'X5', 'X6']]
y = df['Y']
X = sm.add_constant(X)

logit = sm.Logit(y, X).fit()
print(logit.summary2())

pred = (logit.predict(X) > 0.5).astype(int)
print('Accuracy:', accuracy_score(y, pred))
print('Confusion:\n', confusion_matrix(y, pred))

Optimization terminated successfully.
         Current function value: 0.640206
         Iterations 5
                         Results: Logit
Model:              Logit            Method:           MLE      
Dependent Variable: Y                Pseudo R-squared: 0.070    
Date:               2026-05-04 07:04 AIC:              169.3320 
No. Observations:   126              BIC:              180.6771 
Df Model:           3                Log-Likelihood:   -80.666  
Df Residuals:       122              LL-Null:          -86.764  
Converged:          1.0000           LLR p-value:      0.0067395
No. Iterations:     5.0000           Scale:            1.0000   
------------------------------------------------------------------
          Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
------------------------------------------------------------------
const    -3.6276     1.2876   -2.8174   0.0048   -6.1511   -1.1040
X1        0.5805     0.2841    2.0433   0.0410    0.0237    1.1372
X5 

In [45]:
X = df[['X1','X5']]
y = df['Y']
X = sm.add_constant(X)

logit = sm.Logit(y, X).fit()
print(logit.summary2())

pred = (logit.predict(X) > 0.5).astype(int)
print('Accuracy:', accuracy_score(y, pred))
print('Confusion:\n', confusion_matrix(y, pred))

Optimization terminated successfully.
         Current function value: 0.640879
         Iterations 5
                         Results: Logit
Model:              Logit            Method:           MLE      
Dependent Variable: Y                Pseudo R-squared: 0.069    
Date:               2026-05-04 07:23 AIC:              167.5015 
No. Observations:   126              BIC:              176.0103 
Df Model:           2                Log-Likelihood:   -80.751  
Df Residuals:       123              LL-Null:          -86.764  
Converged:          1.0000           LLR p-value:      0.0024455
No. Iterations:     5.0000           Scale:            1.0000   
------------------------------------------------------------------
          Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
------------------------------------------------------------------
const    -3.3848     1.1407   -2.9674   0.0030   -5.6205   -1.1491
X1        0.6199     0.2691    2.3041   0.0212    0.0926    1.1473
X5 

In [43]:
X = df[['X1']]
y = df['Y']
X = sm.add_constant(X)

logit = sm.Logit(y, X).fit()
print(logit.summary2())

pred = (logit.predict(X) > 0.5).astype(int)
print('Accuracy:', accuracy_score(y, pred))
print('Confusion:\n', confusion_matrix(y, pred))

Optimization terminated successfully.
         Current function value: 0.648109
         Iterations 5
                         Results: Logit
Model:              Logit            Method:           MLE      
Dependent Variable: Y                Pseudo R-squared: 0.059    
Date:               2026-05-04 07:16 AIC:              167.3235 
No. Observations:   126              BIC:              172.9960 
Df Model:           1                Log-Likelihood:   -81.662  
Df Residuals:       124              LL-Null:          -86.764  
Converged:          1.0000           LLR p-value:      0.0014006
No. Iterations:     5.0000           Scale:            1.0000   
------------------------------------------------------------------
          Coef.    Std.Err.      z      P>|z|     [0.025    0.975]
------------------------------------------------------------------
const    -3.0901     1.1034   -2.8005   0.0051   -5.2527   -0.9274
X1        0.7574     0.2503    3.0266   0.0025    0.2669    1.2479

Ac

In [41]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=100, random_state=42).fit(X, y)
print('Accuracy:', rf.score(X, y))

Accuracy: 0.6507936507936508


# **Final Summary**

## Final Conclusions

### Best model
Logistic regression on X1 only — 65% accuracy.
Tested: full model (58%), reduced 4-feature (60%), X1-only (65%), Random Forest (65%).

### Why we didn't hit 73%
- Sample size: only 126 customers
- Signal: only X1 (on-time delivery) shows consistent statistical significance across correlation, t-test, and regression
- Multiple models converge at ~65%, indicating the ceiling is the data, not the algorithm

### Bonus — Feature importance for survey reduction

Ranked by combined evidence (correlation, t-test, factor analysis, regression):

| Rank | Feature | Recommendation |
|------|---------|----------------|
| 1 | X1 (on-time delivery) | **Keep** — only consistent predictor across all tests; X1 alone achieves full 65% accuracy |
| 2 | X5 (courier satisfaction) | Drop — significant in t-test but adds no accuracy gain |
| 3 | X6 (app ease) | Drop — borderline t-test, no regression significance |
| 4 | X3 (ordered everything) | Drop — borderline t-test, no regression significance |
| 5 | X4 (good price) | **Drop** — fails correlation, t-test, and regression |
| 6 | X2 (contents as expected) | **Drop** — fails every test; lowest communality (0.10) |

**Minimum viable survey: X1 alone** preserves the full 65% predictive accuracy. All other questions can be removed if the sole goal is predicting customer happiness.
### Recommendation to the business
1. Minimum viable survey: X1 alone — preserves full 65% predictive accuracy.
X2 and X4 should be removed (zero predictive value). X3, X5, X6 add no measurable accuracy gain on this sample and can also be removed if the only goal is happiness prediction.
3. Collect more data — current sample (n=126) is too small for stronger models